app/src/usb_hid.c

```diff
@@ -15,19 +15,73 @@
 #include <zmk/keymap.h>
 #include <zmk/event_manager.h>
 
+#if IS_ENABLED(CONFIG_ZMK_USB_BOOT)
+#include <zmk/events/usb_conn_state_changed.h>
+#endif
+
 LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
 
+#define HID_GET_REPORT_TYPE_MASK 0xff00
+#define HID_GET_REPORT_ID_MASK   0x00ff
+
+#define HID_REPORT_TYPE_INPUT    0x100
+#define HID_REPORT_TYPE_OUTPUT   0x200
+#define HID_REPORT_TYPE_FEATURE  0x300
+
+static int zmk_usb_get_report(const struct device *dev,
+                             struct usb_setup_packet *setup, int32_t *len,
+                             uint8_t **data) {
+
+    /*
+     * 7.2.1 of the HID v1.11 spec is unclear about handling requests for reports that do not exist
+     * For requested reports that aren't input reports, return -ENOTSUP like the Zephyr subsys does
+     */
+    if ((setup->wValue & HID_GET_REPORT_TYPE_MASK) != HID_REPORT_TYPE_INPUT) {
+        LOG_ERR("Unsupported report type %d requested", (setup->wValue & HID_GET_REPORT_TYPE_MASK) << 8);
+        return -ENOTSUP;
+    }
+
+    uint8_t *report = NULL;
+    switch (setup->wValue & HID_GET_REPORT_ID_MASK) {
+        case HID_REPORT_ID_KEYBOARD:
+            report = zmk_hid_get_keyboard_report(HID_REPORT_FULL, HID_PROTOCOL_REPORT);
+            *len = sizeof(struct zmk_hid_keyboard_report);
+            break;
+        case HID_REPORT_ID_CONSUMER:
+            report = zmk_hid_get_consumer_report(HID_REPORT_FULL);
+            *len = sizeof(struct zmk_hid_consumer_report);
```

Comment:
This is making some assumptions about the return values that might easily be changed to not be true. If these functions are going to return byte buffers instead of pointers with known types, then they should really provide the sizes to go along with them. As I mentioned above though, I feel like it would make more sense to just return pointers with types so you can directly get the size from the data.